daemon/analytics.c

```diff
@@ -0,0 +1,195 @@
+    setenv("NETDATA_STOCK_CONFIG_DIR" , verify_required_directory(netdata_configured_stock_config_dir), 1);
+    setenv("NETDATA_PLUGINS_DIR"      , verify_required_directory(netdata_configured_primary_plugins_dir),      1);
+    setenv("NETDATA_WEB_DIR"          , verify_required_directory(netdata_configured_web_dir),          1);
+    setenv("NETDATA_CACHE_DIR"        , verify_required_directory(netdata_configured_cache_dir),        1);
+    setenv("NETDATA_LIB_DIR"          , verify_required_directory(netdata_configured_varlib_dir),       1);
+    setenv("NETDATA_LOCK_DIR"         , netdata_configured_lock_dir, 1);
+    setenv("NETDATA_LOG_DIR"          , verify_required_directory(netdata_configured_log_dir),          1);
+    setenv("HOME"                     , verify_required_directory(netdata_configured_home_dir),         1);
+    setenv("NETDATA_HOST_PREFIX"      , netdata_configured_host_prefix, 1);
+
+    char *default_port = appconfig_get(&netdata_config, CONFIG_SECTION_WEB, "default port", NULL);
+    int clean = 0;
+    if (!default_port) {
+        default_port = strdupz("19999");
+        clean = 1;
+    }
+
+    setenv("NETDATA_LISTEN_PORT"      , default_port, 1);
+    if(clean)
+        freez(default_port);
+
+    get_system_timezone();
+
+    // set the path we need
+    char path[1024 + 1], *p = getenv("PATH");
+    if(!p) p = "/bin:/usr/bin";
+    snprintfz(path, 1024, "%s:%s", p, "/sbin:/usr/sbin:/usr/local/bin:/usr/local/sbin");
+    setenv("PATH", config_get(CONFIG_SECTION_PLUGINS, "PATH environment variable", path), 1);
+
+    // python options
+    p = getenv("PYTHONPATH");
+    if(!p) p = "";
+    setenv("PYTHONPATH", config_get(CONFIG_SECTION_PLUGINS, "PYTHONPATH environment variable", p), 1);
+
+    // disable buffering for python plugins
+    setenv("PYTHONUNBUFFERED", "1", 1);
+
+    // switch to standard locale for plugins
+    setenv("LC_ALL", "C", 1);
+}
+
+void send_statistics( const char *action, const char *action_result, const char *action_data) {
+    static char *as_script;
+
+    if (netdata_anonymous_statistics_enabled == -1) {
+        char *optout_file = mallocz(sizeof(char) * (strlen(netdata_configured_user_config_dir) +strlen(".opt-out-from-anonymous-statistics") + 2));
+        sprintf(optout_file, "%s/%s", netdata_configured_user_config_dir, ".opt-out-from-anonymous-statistics");
+        if (likely(access(optout_file, R_OK) != 0)) {
+            as_script = mallocz(sizeof(char) * (strlen(netdata_configured_primary_plugins_dir) + strlen("anonymous-statistics.sh") + 2));
+            sprintf(as_script, "%s/%s", netdata_configured_primary_plugins_dir, "anonymous-statistics.sh");
+            if (unlikely(access(as_script, R_OK) != 0)) {
+               netdata_anonymous_statistics_enabled=0;
+               info("Anonymous statistics script %s not found.",as_script);
+               freez(as_script);
```

Comment:
This calll for `freez` is not in the best place, please, move outside the conditions.

Reply:
Hi Thiago!

I agree, it is perhaps not the best place, but the `as_script` I think is supposed to be there (defined as `static`) all the time...

On the first call to `send_statistics`, if we don't know if we should send statistics or not (`if (netdata_anonymous_statistics_enabled == -1)`) the variable is setup with the script path, and sets up `netdata_anonymous_statistics_enabled` to either 0 or 1.

On subsequent calls to `send_statistics`, it has already decided that we should send or not, so the `as_script` is not defined/malloced again. If we have freed it after the if, then it will not be available to include in the `command_to_run`.

Ideally it should be freed after the last call to `send_statistics`, I will think about it....

Reply:
I understood your point of view, my unique concern is with the possibility this variable will never be cleaned, because this variable is kept inside the function.

Reply:
Yes, indeed you are right, it will stay there until the agent exits... We'll see if we can up with a certain point perhaps to free it.